tmk_core/common/chibios/eeprom_stm32f4.c

```diff
@@ -0,0 +1,378 @@
+    __IO uint32_t i = 0;
+    for (i = 0xFF; i != 0; i--) {
+        __asm__ volatile("nop\n\t");
+    }
+}
+
+static FLASH_Status FLASH_WaitF4(void) {
+    /* Check for the Flash Status */
+    FLASH_Status status = FLASH_StatusF4();
+    /* Wait for a Flash operation to complete */
+    while( status == FLASH_BUSY) {
+        FLASH_DelayF4();
+        status = FLASH_StatusF4();
+    }
+
+    return status;
+}
+
+static void FLASH_ClearFlagF4(void) {
+    if (FLASH->SR & FLASH_SR_PGAERR) {
+        FLASH->SR |= FLASH_SR_PGAERR;
+    }
+    if (FLASH->SR & FLASH_SR_PGPERR) {
+        FLASH->SR |= FLASH_SR_PGPERR;
+    }
+    if (FLASH->SR & FLASH_SR_PGSERR) {
+        FLASH->SR |= FLASH_SR_PGSERR;
+    }
+    if (FLASH->SR & FLASH_SR_WRPERR) {
+        FLASH->SR |= FLASH_SR_WRPERR;
+    }
+}
+
+static void FLASH_UnlockF4(void) {
+    FLASH->KEYR = FLASH_KEY1_F4;
+    FLASH->KEYR = FLASH_KEY2_F4;
+}
+
+static FLASH_Status FLASH_EraseSectorF4(uint32_t sector) {
+    FLASH_Status status = FLASH_WaitF4();
+    if (status == FLASH_COMPLETE) {
+        /* if the previous operation is completed, proceed to erase the page */
+        FLASH->CR &= ~FLASH_CR_SNB;
+        FLASH->CR |= (sector << FLASH_CR_SNB_POS);
+        FLASH->CR |= FLASH_CR_SER;
+        FLASH->CR |= FLASH_CR_STRT;
+
+        /* Wait for last operation to be completed */
+        status = FLASH_WaitF4();
+        /* clear the erase bit */
+        FLASH->CR &= ~FLASH_CR_SNB;
+        FLASH->CR &= ~FLASH_CR_SER;
+        if (status != FLASH_COMPLETE) {
+            FLASH_ClearFlagF4();
+        }
+    }
+    /* Return the Erase Status */
+    return status;
+}
+
+static FLASH_Status FLASH_ProgramWordF4(uint32_t address, uint32_t data) {
+    FLASH_Status status;
+    if ((address % sizeof(uint32_t) != 0) || !IS_VALID_FLASH(address)) {
+        // not a valid address
+        return FLASH_BAD_ADDRESS;
+    }
+
+    /* Wait for last operation to be completed */
+    status = FLASH_WaitF4();
+    if (status == FLASH_COMPLETE) {
+        // set size to word
+        FLASH->CR &= ~FLASH_CR_PSIZE;
+        FLASH->CR |= FLASH_PSIZE_WORD;
+        FLASH->CR |= FLASH_CR_PG;
+
+        *(__IO uint32_t *)address = data;
+        /* Wait for last operation to be completed */
+        status = FLASH_WaitF4();
+        /* clear the pg bit */
+        FLASH->CR &= ~FLASH_CR_PG;
+        if (status != FLASH_COMPLETE) {
+            FLASH_ClearFlagF4();
+        }
+    }
+    return status;
+}
+
+static uint32_t FLASH_FindValidAddress(void) {
+    uint32_t begin = EE_SECTOR_BEGIN;
+    uint32_t end = EE_SECTOR_END;
+    while( begin < end) {
+        uint32_t data = *(__IO uint32_t*)(begin);
+        if (data == FLASH_EMPTY_VALUE) {
+            return begin;
+        }
+        begin += 4;
+    }
+    return FLASH_EMPTY_VALUE;
+}
+
+static void FLASH_Backup(void) {
+    uint32_t begin = EE_SECTOR_BEGIN;
+    uint32_t end = EE_SECTOR_END;
+    memset(&DataBuf[0], 0xFF, sizeof(DataBuf));
```

Comment:
Again, 0x00 vs. 0xFF.

Reply:
Just confirm that it's writing 0xFF to flash when erasing sectors. Any attempted write with an invalid address will be skipped anyway.